server/src/utils/param_parser.rs

```diff
@@ -0,0 +1,195 @@
+use serde_json::Value;
+
+enum IndexingState {
+    LookingForLeftParenthesis,
+    Accumulating(usize),
+}
+
+// updates a json Value at the specified path
+pub fn update_json_object<'a, V: AsRef<str>>(obj: &'a mut Value, path: &[&str], val: V) {
+    update_json_object_internal(obj, path, val, false)
+}
+
+fn update_json_object_internal<'a, V: AsRef<str>>(
+    obj: &'a mut Value,
+    path: &[&str],
+    val: V,
+    recursion_protection: bool,
+) {
+    if path.len() == 0 {
+        return;
+    }
+    if let Value::Object(the_map) = obj {
+        if path.len() == 1 {
+            the_map.insert(path[0].into(), Value::String(val.as_ref().to_string()));
+        } else {
+            match the_map.get_mut(path[0]) {
+                Some(inner) => {
+                    if inner.is_object() {
+                        // we have a member at the specified index and it is an object (we can insert at path)
+                        update_json_object_internal(inner, &path[1..], val, false);
+                    }
+                }
+                None => {
+                    if recursion_protection {
+                        //this is a bug we should NEVER be here
+                        log::error!("update_value_internal, infinite recursion detected");
+                    } else {
+                        //nothing yet at the specified path create an object
+                        the_map.insert(path[0].into(), Value::Object(serde_json::Map::new()));
+                        // now we should have an object at the path, try again
+                        update_json_object_internal(obj, path, val, true);
+                    }
+                }
+            }
+        }
+    }
+}
+
+/// Merge two serde values
+/// taken (with small changes) from stack overflow answer
+/// https://stackoverflow.com/questions/47070876/how-can-i-merge-two-json-objects-with-rust
+pub fn merge_vals(a: &mut Value, b: Value) {
+    match (a, b) {
+        //recursively merge dicts
+        (a @ &mut Value::Object(_), Value::Object(b)) => {
+            let a = a.as_object_mut().unwrap();
```

Comment:
Is there a reason this does not match the inner map right away?

```rust
(&mut Value::Object(ref mut a), Value::Object(b)) => { ... }
```

Reply:
The suggested code does not compile:
```
error[E0009]: cannot bind by-move and by-ref in the same pattern
  --> server/src/utils/param_parser.rs:41:55
   |
   |         (&mut Value::Object(ref mut a), Value::Object(b)) => {
   |                             ---------                 ^ by-move pattern here
   |                             |
   |                             both by-ref and by-move used

```

Couldn't figure out how to write it close to what you are suggesting.
Let's have a closer look together.

Reply:
Ah, that makes sense. Let's leave it then 👍